From: =?utf-8?q?Christian_G=C3=B6ttsche?= <cgzones@googlemail.com>
Date: Fri, 17 Jan 2025 18:25:17 +0100
Applied-Upstream: https://github.com/jvdmr/mod_evasive/commit/14c1d50b37ab2038357156f422361fb49f03b04b
Subject: Drop outdated hashtable entries

Drop IP entries that had no hits in more than 6 hours.
---
 mod_evasive24.c | 27 ++++++++++++++++++++++++---
 1 file changed, 24 insertions(+), 3 deletions(-)

diff --git a/mod_evasive24.c b/mod_evasive24.c
index 5288a09..c64efb1 100644
--- a/mod_evasive24.c
+++ b/mod_evasive24.c
@@ -759,6 +759,12 @@ static struct ntt_node *ntt_find(struct ntt *ntt, const char *key) {
     return((struct ntt_node *)NULL);
 }
 
+/* Whether a node is outdated */
+
+static int ntt_node_is_outdated(const struct ntt_node *node, apr_time_t timestamp) {
+    return timestamp - node->timestamp >= 6 * 60 * 60; /* 6 hours */
+}
+
 /* Insert a node into the tree */
 
 static struct ntt_node *ntt_insert(struct ntt *ntt, const char *key, apr_time_t timestamp) {
@@ -777,12 +783,27 @@ static struct ntt_node *ntt_insert(struct ntt *ntt, const char *key, apr_time_t
         if (strcmp(key, node->key) == 0) {
             new_node = node;
             node = NULL;
+            break;
         }
 
-        if (new_node == NULL) {
-            parent = node;
-            node = node->next;
+        /* Delete outdated entries */
+        if (ntt_node_is_outdated(node, timestamp)) {
+            struct ntt_node *next = node->next;
+
+            if (parent)
+                parent->next = next;
+            else
+                ntt->tbl[hash_code] = next;
+
+            free(node->key);
+            free(node);
+            ntt->items--;
+            node = next;
+            continue;
         }
+
+        parent = node;
+        node = node->next;
     }
 
     if (new_node != NULL) {
